### PR TITLE
Include additional Container[*].SecurityContext properties

### PR DESCRIPTION
**Pull request by dprotaso**
tl;dr

`runAsGroup` - REQUIRED, if securityContext is supported. This compliments `runAsUser`
`readOnlyRootFilesystem` - REQUIRED, if securityContext is supported.

Including `readOnlyRootFilesystem` for the purposes of reserving the keyword.

### CODE REVIEW

**Comment by dprotaso** [comment]
cc @knative/trademark-committee

**Comment by duglin** [comment]
Just curious, why didn't you make readOnlyRootFilesystem required?

**Comment by dprotaso** [comment]
I don't have strong opinions - I can change if the KTC feels that way

In contrast I felt like `runAsGroup` compliments `runAsUser` thus it has more utility and should be `REQUIRED`. Meaning if people are overriding UIDs they're probably also overriding GIDs

**Comment by duglin** [comment]
Given a choice I'd vote for making as many things as possible "REQUIRED to support" as that'll only help interop. But I'd be curious to know if other TMC members have a concern.

**Comment by omerbensaadon** [comment]
Gonna leave this one for a few days so @spencerdillard can get more info from CloudRun folks

File: specs/serving/knative-api-specification-1.0.md [hunk 1]
```diff
@@ -2215,6 +2215,26 @@ Max: 1
    <td><strong>Schema Requirement</strong>
    </td>
   </tr>
+  <tr>
+   <td><code>readOnlyRootFilesystem</code>
```
Comment (rhuss) [inline specs/serving/knative-api-specification-1.0.md:2219]: In https://github.com/knative/specs/pull/73/files#L1523 it is mentioned that "only runAsUser may be specified". This is in contracdictionon to this section.

**Comment by rhuss** [comment]
> Including `readOnlyRootFilesystem` for the purposes of reserving the keyword.

Any particular reason why only "reserving" this key word and not the other parts of a SecurityContext that are not yet supported, like "capabilities" or "allowPrivilegeEscalation" ? I don't see really why "readOnlyRootFilesystem" is 'special' here.

I'm not so deep in the conformance discussions but is "RECOMMENDED" the standard way to include reserved, potentially nowhere implemented keywords. Wouldn't "MAY" be a better fit as I would expect with "RECOMMENDED" that at least one current implementation would support it ?

**Comment by dprotaso** [comment]
bump @spencerdillard @omerbensaadon

File: specs/serving/knative-api-specification-1.0.md [hunk 2]
```diff
@@ -2215,6 +2215,26 @@ Max: 1
...
+   </td>
+   <td>boolean
+   </td>
+   <td>As specified in Kubernetes <a href="https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#securitycontext-v1-core">core/v1.SecurityContext</a>
+   </td>
+   <td>RECOMMENDED, if securityContext is supported.
```
Comment (evankanderson) [inline specs/serving/knative-api-specification-1.0.md:2225]: Note that this is part of the [PodSecurityStandards for the `restricted` level](https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted)

I think Dave Bailey and Greg Block at Google were interested this back circa 2017, but there was no representation of this property in Knative at the time. Note that this aligns with the runtime spec here: https://github.com/knative/specs/blob/main/specs/serving/runtime-contract.md#root

File: specs/serving/knative-api-specification-1.0.md [hunk 3]
```diff
@@ -2215,6 +2215,26 @@ Max: 1
...
+  <tr>
+   <td><code>runAsGroup</code>
+   </td>
+   <td>int
+   </td>
+   <td>As specified in Kubernetes <a href="https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#securitycontext-v1-core">core/v1.SecurityContext</a>
+   </td>
+   <td>REQUIRED, if securityContext is supported.
+   </td>
+  </tr>
```
Comment (evankanderson) [inline specs/serving/knative-api-specification-1.0.md:2237]: Ditto, in terms of [PodSecurityStandards](https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted).

We probably also need to update the [Runtime contract](https://github.com/knative/specs/blob/main/specs/serving/runtime-contract.md#user).

**Review by evankanderson — COMMENTED** [review]
/hold

I'd like to get someone from Google to look at these vis-a-vis Cloud Run, but these generally seem attractive in terms of allowing users to be able to lock down their execution environment more than the Kubernetes default.

More generally, I'd be sympathetic to any spec changes that might be needed to allow a conformant implementation to run on a Kubernetes namespace which is running in the [`restricted` PodSecurity policy](https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted).

**Comment by evankanderson** [comment]
This needs to update the runtime contract as well to specify the group when `runAsGroup` is set and when `readOnlyRootFilesystem` is set: https://github.com/knative/specs/blob/main/specs/serving/runtime-contract.md#user and https://github.com/knative/specs/blob/main/specs/serving/runtime-contract.md#default-filesystems

**Comment by spencerdillard** [comment]
Having had some time to think about this and discuss with others, it seems to me that runAsGroup should be required in order to avoid accidentally having some default group that is dangerous, so ok with that part. But the readOnlyFileSystem seems like it should be recommended, since it is a boolean. As long as it is clear that the default is false and it is rw, then making it required seems like an unnecessary burden. It should only be used when the user doesn't want the default value.
Open to other takes, of course.

**Comment by duglin** [comment]
@spencerdillard  someone will correct me if I'm wrong, but I think the "recommended" vs "required" here is about support for the field, not whether the user needs to specify a value.

**Comment by spencerdillard** [comment]
True -- I do think I got a bit hyperfocused on the user experience there. Put another way, we are making it required that any implementation must support ROFS or will not be conformant. So, I agree that this is reasonable to require. Sorry for getting off track there ;-)

/lgtm

**Comment by dprotaso** [comment]
@spencerdillard you'll need to `/approve` without the `d`

**Comment by dprotaso** [comment]
or any other @knative/trademark-committee members

**Comment by rhuss** [comment]
/lgtm

however:

> `readOnlyRootFilesystem` - RECOMMENDED, if securityContext is supported.

The tl;dr from the description does not fit the actual PR where support for `readOnlyRootFileSystem` is set to [REQUIRED](https://github.com/knative/specs/pull/73/files#diff-aa3e1d27ae58f8a946c5e2cae6b6e57a22109d58934462f5608bd6cbf0353584R2225)

**Comment by dprotaso** [comment]
I see `readOnlyRootFilesystem` as `REQUIRED` with Evan's changes that merged into my branch

see:
https://github.com/knative/specs/blob/15121935c0cf2acaee97a02f87d3a17843ae07ae/specs/serving/knative-api-specification-1.0.md#securitycontext

**Comment by rhuss** [comment]
Yep, I updated the description of this PR and changed RECOMMENDED to REQUIRED.

**Pull request merged**